File: 搜索与回溯/113.py
from typing import List
# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def pathSum(self, root: TreeNode, target: int) -> List[List[int]]:
        def dfs(root, target, res, path):
            if not root:
                return
            if not root.left and not root.right:
                if sum(path)+root.val == target:
                    res.append(path+[root.val])
                return
            dfs(root.left, target, res, path+[root.val])
            dfs(root.right, target, res, path+[root.val])
        res = []
        dfs(root, target, res, [])
        return res


def stringToTreeNode(input):
    input = input.strip()
    input = input[1:-1]
    if not input:
        return None

    inputValues = [s.strip() for s in input.split(',')]
    root = TreeNode(int(inputValues[0]))
    nodeQueue = [root]
    front = 0
    index = 1
    while index < len(inputValues):
        node = nodeQueue[front]
        front = front + 1

        item = inputValues[index]
        index = index + 1
        if item != "null":
            leftNumber = int(item)
            node.left = TreeNode(leftNumber)
            nodeQueue.append(node.left)

        if index >= len(inputValues):
            break

        item = inputValues[index]
        index = index + 1
        if item != "null":
            rightNumber = int(item)
            node.right = TreeNode(rightNumber)
            nodeQueue.append(node.right)
    return root


root = stringToTreeNode("[5,4,8,11,null,13,4,7,2,null,null,null,1]")
targetSum = 22
obj = Solution()
print(obj.pathSum(root, targetSum))

File: 搜索与回溯/113.cpp
#include <vector>
#include <numeric>

using namespace std;
//Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<vector<int>> pathSum(TreeNode *root, int targetSum)
    {
        vector<vector<int>> res;
        vector<int>path;
        dfs(root,targetSum,res,path);
        return res;
    }

    void dfs(TreeNode *root, int targetSum, vector<vector<int>> &res, vector<int> &path)
    {
        if (root){
            path.push_back(root->val);
            targetSum -= root->val;
            if (!root->left && !root->right && targetSum==0){
                res.push_back(path);
            }
            dfs(root->left,targetSum,res,path);
            dfs(root->right,targetSum,res,path);
            path.pop_back();
        }
    }
};